File: fashionmix-studio/backend/rule_scorer.py
"""
Pure-Python rule-based outfit scorer.

Mirrors frontend/lib/rule-scorer.js (must stay in sync).
Spec: docs/superpowers/specs/2026-06-27-fashionmix-studio-v02-design.md §5.4
"""
from __future__ import annotations

from typing import Any


class ScorerError(ValueError):
    """Raised when input item is missing required fields."""


REQUIRED_FIELDS = ("id", "category", "slot", "price", "styleTags", "riskTags")


def _validate(items: list[dict[str, Any]]) -> None:
    for it in items:
        missing = [f for f in REQUIRED_FIELDS if f not in it]
        if missing:
            raise ScorerError(f"item {it.get('id', '?')!r} missing fields: {missing}")


def _layer_completeness(items: list[dict[str, Any]]) -> int:
    slots = {it["slot"] for it in items}
    has = lambda s: s in slots
    score = 0
    if has("upper"):
        score += 30
    if has("lower"):
        score += 30
    if has("feet"):
        score += 25
    if has("neck") or has("extra") or has("head"):
        score += 15
    return min(score, 100)


def _style_consistency(items: list[dict[str, Any]]) -> int:
    if len(items) < 2:
        return 0
    sets = [set(it["styleTags"]) for it in items]
    intersection = set.intersection(*sets) if sets else set()
    union = set.union(*sets) if sets else set()
    if not union or not intersection:
        return 0
    raw = (len(intersection) ** 2) / (len(union) * len(intersection))
    return min(int(raw * 100), 100)


def _color_harmony(items: list[dict[str, Any]]) -> int:
    distinct_colors = {t for it in items for t in it["styleTags"] if t.endswith("系")}
    n = len(distinct_colors)
    if n <= 1:
        return 100
    if n == 2:
        return 80
    if n == 3:
        return 60
    return 40


def _weighted_avg(items: list[dict[str, Any]], key: str) -> int:
    total_price = sum(it["price"] for it in items) or 1
    weighted = sum(it["price"] * it.get(key, 0) for it in items)
    return int(weighted / total_price)


def _risk_score(items: list[dict[str, Any]]) -> int:
    total_risks = sum(len(it["riskTags"]) for it in items)
    return max(0, 100 - total_risks * 15)


def _collect_tags(items: list[dict[str, Any]], key: str) -> list[str]:
    seen: list[str] = []
    for it in items:
        for t in it[key]:
            if t not in seen:
                seen.append(t)
    return seen


def score_outfit(items: list[dict[str, Any]]) -> dict[str, Any]:
    """Score an outfit purely from item attributes. No LLM involved."""
    _validate(items)
    return {
        "scores": {
            "styleConsistency": _style_consistency(items),
            "colorHarmony": _color_harmony(items),
            "layerCompleteness": _layer_completeness(items),
            "photoScore": _weighted_avg(items, "photoScore"),
            "dailyScore": _weighted_avg(items, "dailyScore"),
            "riskScore": _risk_score(items),
        },
        "styleTags": _collect_tags(items, "styleTags"),
        "riskTags": _collect_tags(items, "riskTags"),
        "suggestion": "规则评分：未调用 AI。",
        "source": "rule-fallback",
    }

File: fashionmix-studio/frontend/lib/rule-scorer.js
/**
 * Pure-JS mirror of backend/rule_scorer.py (must stay in sync).
 * Spec: docs/superpowers/specs/2026-06-27-fashionmix-studio-v02-design.md §5.4
 */

export class ScorerError extends Error {}

const REQUIRED_FIELDS = ['id', 'category', 'slot', 'price', 'styleTags', 'riskTags'];

function validate(items) {
  for (const it of items) {
    const missing = REQUIRED_FIELDS.filter(f => !(f in it));
    if (missing.length) {
      throw new ScorerError(`item ${it.id || '?'} missing: ${missing.join(', ')}`);
    }
  }
}

function layerCompleteness(items) {
  const slots = new Set(items.map(it => it.slot));
  let s = 0;
  if (slots.has('upper')) s += 30;
  if (slots.has('lower')) s += 30;
  if (slots.has('feet')) s += 25;
  if (slots.has('neck') || slots.has('extra') || slots.has('head')) s += 15;
  return Math.min(s, 100);
}

function styleConsistency(items) {
  if (items.length < 2) return 0;
  const sets = items.map(it => new Set(it.styleTags));
  const inter = [...sets[0]].filter(x => sets.every(s => s.has(x)));
  const union = new Set(items.flatMap(it => it.styleTags));
  if (!union.size || !inter.length) return 0;
  const raw = (inter.length ** 2) / (union.size * inter.length);
  return Math.min(Math.round(raw * 100), 100);
}

function colorHarmony(items) {
  const distinct = new Set(items.flatMap(it => it.styleTags.filter(t => t.endsWith('系'))));
  const n = distinct.size;
  if (n <= 1) return 100;
  if (n === 2) return 80;
  if (n === 3) return 60;
  return 40;
}

function weightedAvg(items, key) {
  const total = items.reduce((a, it) => a + it.price, 0) || 1;
  const w = items.reduce((a, it) => a + it.price * (it[key] || 0), 0);
  return Math.round(w / total);
}

function riskScore(items) {
  const total = items.reduce((a, it) => a + it.riskTags.length, 0);
  return Math.max(0, 100 - total * 15);
}

function collectTags(items, key) {
  const seen = new Set();
  const out = [];
  for (const it of items) {
    for (const t of it[key]) {
      if (!seen.has(t)) { seen.add(t); out.push(t); }
    }
  }
  return out;
}

export function scoreOutfit(items) {
  validate(items);
  return {
    scores: {
      styleConsistency: styleConsistency(items),
      colorHarmony: colorHarmony(items),
      layerCompleteness: layerCompleteness(items),
      photoScore: weightedAvg(items, 'photoScore'),
      dailyScore: weightedAvg(items, 'dailyScore'),
      riskScore: riskScore(items),
    },
    styleTags: collectTags(items, 'styleTags'),
    riskTags: collectTags(items, 'riskTags'),
    suggestion: '规则评分：未调用 AI。',
    source: 'rule-fallback',
  };
}
